File: rabbit-mq/work-queues/worker.py
#!/usr/bin/env python
import pika
import time
import sys
import os


def main():
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host='localhost'))
    channel = connection.channel()

    # durable = True sets message durability
    # auto_ack=False, this will ensure message acknoledgement is setup -> no messages missed when worker is shutdown
    channel.queue_declare(queue='task_queue', durable=True)
    print(' [*] Waiting for messages. To exit press CTRL+C')

    def callback(ch, method, properties, body):
        print(" [x] Received %r" % body.decode())
        task_time = body.count(b'.')
        time.sleep(task_time)
        print(" [x] Done %d seconds" % task_time)
        # It's time to remove the auto_ack flag and send a proper acknowledgment from the worker, once we're done with a task.
        ch.basic_ack(delivery_tag=method.delivery_tag)

    # fair queue prefetch_count = 1
    channel.basic_qos(prefetch_count=1)
    channel.basic_consume(queue='task_queue', on_message_callback=callback)

    channel.start_consuming()


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('Interrupted')
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)

File: rabbit-mq/work-queues/worker.jl
using AMQPClient

Base.exit_on_sigint(false)

abstract type AbstractConnection end

struct Connection <: AbstractConnection
    port 
    login
    password
    conn
    chan
end

function connect()
    # Establish the connection to the RabbitMQ Server
    port = AMQPClient.AMQP_DEFAULT_PORT
    login = "guest"  # default is usually "guest"
    password = "guest"  # default is usually "guest"
    auth_params = Dict{String,Any}("MECHANISM" => "AMQPLAIN", "LOGIN" => login, "PASSWORD" => password)

    conn = connection(; virtualhost="/", host="localhost", port=port, auth_params=auth_params, amqps=nothing)

    chan = channel(conn, AMQPClient.UNUSED_CHANNEL, true)
    
    return Connection(port, login, password, conn, chan)
end


function main(connection::AbstractConnection)
    chan = connection.chan

    success, message_count, consumer_count = queue_declare(chan, "task_queue", durable=true)
    
    println(" [*] Waiting for messages. To exit press CTRL+C")

    callback = rcvd_msg -> begin
        msg_str = String(rcvd_msg.data)
        println(" [x] Received '$msg_str'")
        task_time = sum([1 for i = eachmatch(r".", msg_str)])
        sleep(task_time)
        println(" [x] Done $task_time seconds")
        # It's time to remove the auto_ack flag and 
        # send a proper acknowledgment from the worker, 
        # once we're done with a task.
        basic_ack(chan, rcvd_msg.delivery_tag)
    end
    
    # Define qos parameters
    prefetch_size = 0
    prefetch_count = 1
    global_qos = false
    basic_qos(chan, prefetch_size, prefetch_count, global_qos)
    
    success, consumer_tag = basic_consume(chan, "task_queue", callback)
    
    @assert success
    # println("consumer registered with tag $consumer_tag")

    # go ahead with other stuff...
    # or wait for an indicator for shutdown

    while true
        sleep(1)
    end

end


try
    connection = connect()
    main(connection)
catch e
    if e isa InterruptException
        println("Exited")
        # unsubscribe the consumer from the queue
        # basic_cancel(connection, consumer_tag)
    else
        print("There was an error")
        print(e)
    end
    exit()
end

